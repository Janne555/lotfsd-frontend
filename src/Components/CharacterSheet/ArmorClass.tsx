import React from 'react'
import { createUseStyles } from 'react-jss'
import Cube from '../_shared/Cube'
import Triangle from '../_shared/Triangle'

const useStyles = createUseStyles((theme: Theme) => ({
  armorClassRoot: {
    display: 'grid',
    gridTemplateColumns: 'auto auto',
    gridTemplateRows: 'auto auto',
    gridColumnGap: 10,
    gridRowGap: 10,
    justifyItems: 'center',
    border: theme.border,
    padding: theme.padding,
    textAlign: 'center',
    '& h4:nth-of-type(1)': {
      gridRowStart: 2
    },
    '& h4:nth-of-type(3)': {
      gridRowStart: 4
    }
  }
}))


function ArmorClass() {
  const classes = useStyles()

  return (
    <div className={classes.armorClassRoot}>
      <Cube>1</Cube>
      <h4>Mêlée</h4>
      <Cube>1</Cube>
      <h4>Ranged</h4>
      <Triangle>2</Triangle>
      <h4>Without<br />Shield</h4>
      <Triangle>2</Triangle>
      <h4>Surprised</h4>
    </div>
  )
}

export default ArmorClass