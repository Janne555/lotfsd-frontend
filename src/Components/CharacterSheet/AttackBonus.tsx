import React from 'react'
import { useSelector, useCharacterContext, useDispatch } from '../../hooks'
import { createUseStyles } from 'react-jss'
import RotatedCube from '../_shared/RotatedCube'
import { selectBaseAttackBonus, selectMeleeAttackBonus, selectRangedAttackBonus } from '../../Redux/selectors'
import { TITLES } from '../../constants'
import Input from '../_shared/Input'
import { Validator } from '../../services'
import { setAttackBonus } from '../../Redux/actions'

const useStyles = createUseStyles((theme: Theme) => ({
  attackBonus: {
    display: 'grid',
    gridTemplateColumns: 'repeat(3, 75px)',
    gridTemplateRows: 'auto 40px auto',
    gridRowGap: 20
  },
  title: {
    gridColumnStart: 1,
    gridColumnEnd: 4,
    justifySelf: 'center'
  }
}))

function AttackBonus() {
  const classes = useStyles()
  const { characterId } = useCharacterContext()
  const baseAB = useSelector(selectBaseAttackBonus(characterId))
  const meleeAB = useSelector(selectMeleeAttackBonus(characterId))
  const rangedAB = useSelector(selectRangedAttackBonus(characterId))

  return (
    <div className={classes.attackBonus}>
      <h2 className={classes.title}>Attack Bonus</h2>
      <SingleBonus index={1} value={baseAB} title={TITLES.BASE_AB} base />
      <SingleBonus index={2} value={meleeAB} title={TITLES.MELEE_AB} />
      <SingleBonus index={3} value={rangedAB} title={TITLES.RANGED_AB} />
    </div>
  )
}

const useSubStyles = createUseStyles({
  singleBonus: {
    gridColumnStart: (index: number) => index,
    justifySelf: 'center',
    gridRowStart: 2
  },
  title: {
    gridColumnStart: (index: number) => index,
    justifySelf: 'center',
    gridRowStart: 3
  }
})

const validator = new Validator().isLengthy.isNumber

type SubProps = {
  index: number
  value: number
  title: string
  base?: boolean
}

function SingleBonus({ index, value, title, base }: SubProps) {
  const classes = useSubStyles(index)
  const dispatch = useDispatch()
  const { characterId } = useCharacterContext()

  function handleChange(value: string) {
    dispatch(setAttackBonus({ value: Number(value), characterId }))
  }


  return (
    <>
      <div className={classes.singleBonus}>
        <RotatedCube>
          {base
            ? <Input isValid={validator.validate} onChange={handleChange} value={`${value}`} />
            : value
          }
        </RotatedCube>
      </div>
      <h4 className={classes.title}>{title}</h4>
    </>
  )
}

export default AttackBonus