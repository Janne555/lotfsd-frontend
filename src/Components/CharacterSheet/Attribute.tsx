
import React from 'react'
import { createUseStyles } from 'react-jss'
import { ATTRIBUTE_DETAILS, ATTRIBUTE_TITLES, CHARACTER_SHEET_UPDATE_MUTATION } from '../../constants'
import Input from '../_shared/Input'
import { Validator, calculateModifier, updateCharacterSheet } from '../../services'
import { useMutation } from '@apollo/react-hooks'
import { CharacterSheetUpdateMutation, CharacterSheetUpdateMutationVariables } from '../../constants/mutations/__generated__/CharacterSheetUpdateMutation'

const useAttributeStyles = createUseStyles((theme: Theme) => ({
  title: {
    gridColumnStart: 2,
    gridRowStart: (index: number) => (index + 1) * 2 - 1,
    justifySelf: 'center',
    alignSelf: 'end'
  },
  scoreRoot: {
    border: theme.border,
    gridRowStart: (index: number) => (index + 1) * 2,
    gridColumnStart: 2,
    justifySelf: 'center',
    width: 40,
    height: 40,
    display: 'flex',
    justifyContent: 'center'
  },
  modifierRoot: {
    gridRowStart: (index: number) => (index + 1) * 2,
    gridColumnStart: 3,
    border: theme.border,
    borderRadius: '100%',
    margin: 1,
    display: 'flex',
    justifyContent: 'center',
    width: 40,
    height: 40
  },
  modifier: {
    alignSelf: 'center'
  },
  details: {
    gridRowStart: (index: number) => (index + 1) * 2,
    gridColumnStart: 4
  }
}))

type AttributeProps = {
  title: keyof Attributes
  score: number
  index: number
  modifier?: number
  onChange?: (key: keyof Attributes, value: string) => void
  characterId?: string
}

const validator = new Validator().isLengthy.isNumber

function Attribute({ title, score, index, modifier, onChange, characterId }: AttributeProps) {
  const classes = useAttributeStyles(index)
  const [mutate, { data, loading, error }] = useMutation<CharacterSheetUpdateMutation, CharacterSheetUpdateMutationVariables>(CHARACTER_SHEET_UPDATE_MUTATION, {
    update: updateCharacterSheet(characterId)
  })

  function handleBlur(value: string, isValid: boolean) {
    if (characterId && score !== Number(value) && isValid) {
      mutate({
        variables: {
          ch: { [title]: Number(value) },
          id: characterId
        }
      })
    }
  }

  return (
    <>
      <label htmlFor={title} className={classes.title}>{ATTRIBUTE_TITLES[title]}</label>
      <div className={classes.scoreRoot}>
        <Input
          validate={validator.validate}
          onChange={value => onChange?.(title, value)}
          value={`${score}`}
          id={title}
          type="number"
          disabled={loading}
          onBlur={handleBlur}
        />
      </div>
      <div className={classes.modifierRoot}>
        <span className={classes.modifier}>{modifier != null ? modifier : calculateModifier(score)}</span>
      </div>
      <span className={classes.details}>{ATTRIBUTE_DETAILS[title]}</span>
    </>
  )
}

export default Attribute