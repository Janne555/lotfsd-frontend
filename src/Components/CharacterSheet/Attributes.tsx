import React from 'react'
import { createUseStyles } from 'react-jss'
import { useSelector } from '../../hooks/redux'
import { selectAttributes } from '../../Redux/selectors'
import { ATTRIBUTE_DETAILS } from '../../constants'
import { calculateModifier } from '../../services'

const useStyles = createUseStyles((theme: Theme) => ({
  root: {
    display: 'flex',
    flexDirection: 'column'
  }
}))

export default function Attributes() {
  const attributes = useSelector<Attributes>(selectAttributes)
  const classes = useStyles()

  return (
    <div className={classes.root}>
      {
        Object.entries(attributes).map(([name, value]) => (
          <Attribute key={name} title={name} score={value} />
        ))
      }
    </div>
  )
}

const useAttributeStyles = createUseStyles((theme: Theme) => ({
  root: {
    display: 'grid',
    gridTemplateColumns: '40px 40px 40px auto',
    gridColumnGap: 10
  },
  title: {
    gridColumnStart: 2,
    justifySelf: 'center'
  },
  scoreRoot: {
    border: theme.border,
    gridRowStart: 2,
    gridColumnStart: 2,
    justifySelf: 'center',
    width: 40,
    height: 40,
    display: 'flex',
    justifyContent: 'center'
  },
  score: {
    alignSelf: 'center'
  },
  modifierRoot: {
    gridRowStart: 2,
    gridColumnStart: 3,
    border: theme.border,
    borderRadius: '100%',
    margin: 1,
    display: 'flex',
    justifyContent: 'center',
    width: 40,
    height: 40
  },
  modifier: {
    alignSelf: 'center'
  },
  details: {
    gridRowStart: 2,
    gridColumnStart: 4
  }
}))

type AttributeProps = {
  title: string
  score: number
}

function Attribute({ title, score }: AttributeProps) {
  const classes = useAttributeStyles()

  return (
    <div className={classes.root}>
      <label className={classes.title}>{title}</label>
      <div className={classes.scoreRoot}>
        <span className={classes.score}>{score}</span>
      </div>
      <div className={classes.modifierRoot}>
        <span className={classes.modifier}>{calculateModifier(score)}</span>
      </div>
      <span className={classes.details}>{ATTRIBUTE_DETAILS[title]}</span>
    </div>
  )
}