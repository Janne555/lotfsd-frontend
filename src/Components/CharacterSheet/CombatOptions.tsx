import React from 'react'
import { useSelector } from '../../hooks'
import { selectCombatOptions } from '../../Redux/selectors'
import { createUseStyles } from 'react-jss'

const useStyles = createUseStyles((theme: Theme) => ({
  combatOptions: {
    padding: theme.padding,
    border: theme.border,
    maxWidth: 300,
    '& h2': {
      textAlign: 'center'
    }
  }
}))

export default function CombatOptions() {
  const classes = useStyles()
  const { defensive, parry, press, standard } = useSelector(selectCombatOptions)

  return (
    <div className={classes.combatOptions}>
      <h2>Combat Options</h2>
      
    </div>
  )
}
