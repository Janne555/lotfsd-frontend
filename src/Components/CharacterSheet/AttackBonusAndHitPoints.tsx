import React from 'react'
import { createUseStyles } from 'react-jss'
import AttackBonus from './AttackBonus'
import DieFace from '../_shared/DieFace'
import Cube from '../_shared/Cube'
import { Validator } from '../../services'
import { useCharacterSheetUpdate } from '../../hooks'
import Input from '../_shared/Input'

const useStyles = createUseStyles((theme: Theme) => ({
  attackBonusAndHitPointsRoot: {
    display: 'flex',
    border: theme.border,
    padding: theme.padding,
    flexWrap: 'wrap',
    justifyContent: 'space-around',
    alignItems: 'center',
    backgroundColor: theme.componentBackgroundColor
  },
  surpriseChance: {
    marginTop: 10,
    display: 'grid',
    flexDirection: 'column',
    '& h3': {
      alignSelf: 'end'
    }
  },
  surpriseValue: {
    justifySelf: 'center',
  },
  hitpoints: {
    marginTop: 10,
    display: 'grid',
    gridRowGap: 10,
    justifyItems: 'center'
  }
}))

type Props = {
  baseAB: number
  meleeAB: number
  rangedAB: number
  maxHp: number
  currentHp: number
  surpriseChance: number
  characterId?: string
}

const validator = new Validator().isLengthy.isNumber

function AttackBonusAndHitPoints({ baseAB, currentHp, maxHp, meleeAB, rangedAB, surpriseChance, characterId }: Props) {
  const classes = useStyles()
  const [mutate, { data, loading, error }] = useCharacterSheetUpdate(characterId)

  function handleSurpriseChange(value: number) {
    if (characterId && value !== surpriseChance) {
      mutate({
        variables: {
          ch: { surpriseChance: value },
          id: characterId
        }
      })
    }
  }

  function handleMaxHpBlur(value: string, isValid: boolean) {
    if (characterId && Number(value) !== maxHp && isValid) {
      mutate({
        variables: {
          ch: { maxHp: Number(value) },
          id: characterId
        }
      })
    }
  }

  function handleCurrentHpBlur(value: string, isValid: boolean) {
    if (characterId && Number(value) !== maxHp && isValid) {
      mutate({
        variables: {
          ch: { currentHp: Number(value) },
          id: characterId
        }
      })
    }
  }

  return (
    <div className={classes.attackBonusAndHitPointsRoot}>
      <AttackBonus baseAB={baseAB} meleeAB={meleeAB} rangedAB={rangedAB} characterId={characterId} />
      <div className={classes.surpriseChance}>
        <h3>Surprise Chance</h3>
        <div className={classes.surpriseValue}>
          <DieFace loading={loading} value={surpriseChance} onValueChange={handleSurpriseChange} />
        </div>
      </div>
      <div className={classes.hitpoints}>
        <h4>Max HP</h4>
        <Cube>
          <Input
            validate={validator.validate}
            value={`${maxHp}`}
            type="number"
            disabled={loading}
            onBlur={handleMaxHpBlur}
          />
        </Cube>
        <Cube>
          <Input
            validate={validator.validate}
            value={`${currentHp}`}
            type="number"
            disabled={loading}
            onBlur={handleCurrentHpBlur}
          />
        </Cube>
        <h4>Current HP</h4>
      </div>
    </div>
  )
}

export default AttackBonusAndHitPoints