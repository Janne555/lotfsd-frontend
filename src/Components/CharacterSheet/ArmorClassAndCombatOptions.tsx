import React from 'react'
import ArmorClass from './ArmorClass'
import CombatOptions from './CombatOptions'
import { createUseStyles } from 'react-jss'

const useStyles = createUseStyles((theme: Theme) => ({
  combatOptionsAndAC: {
    display: 'grid',
    gridTemplateColumns: '50% 50%',
    gridColumnGap: '1rem',
    width: 'calc(100% - 1rem)'
  },
  '@media (max-width: 1100px)': {
    combatOptionsAndAC: {
      display: 'block',
      width: '100%',
      '& > div:first-child': {
        marginBottom: '1rem'
      }
    }
  }
}))


export default function ArmorClassAndCombatOptions() {
  const classes = useStyles()
  return (
    <div className={classes.combatOptionsAndAC}>
      <ArmorClass />
      <CombatOptions />
    </div>
  )
}
