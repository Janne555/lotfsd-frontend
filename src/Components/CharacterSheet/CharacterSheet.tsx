import React from 'react'
import Attributes from './Attributes'
import SavingThrows from './SavingThrows'
import AttackBonusAndHitPoints from './AttackBonusAndHitPoints'
import ArmorClassAndCombatOptions from './ArmorClassAndCombatOptions'
import CommonActivities from './CommonActivities'
import EquipmentList from './EquipmentList'
import Encumbrance from './Encumbrance'
import Retainers from './Retainers'
import { createUseStyles } from 'react-jss'
import Languages from './Languages'
import InfoBar from './InfoBar'
import { useSelector } from '../../hooks'
import { selectAttributes, selectAttributeModifiers } from '../../Redux/selectors'
import { Redirect } from 'react-router-dom'
import SpeedDial from '../Interface/SpeedDial'

const useStyles = createUseStyles((theme: Theme) => ({
  characterSheet: {
    backgroundColor: theme.colorGreyLightest,
    padding: '1rem',
    columnCount: 2,
    '& > *': {
      breakInside: 'avoid-column',
      marginBottom: '1rem'
    }
  },
  '@media (max-width: 1100px)': {
    characterSheet: {
      display: 'flex',
      flexDirection: 'column',
      columnCount: 'unset',
      padding: 1
    }
  }
}))

const context = React.createContext({
  characterName: "default",
  characterId: 'default'
})

type Props = {
  characterId: string
  characterName: string
}

function CharacterSheet({ characterId, characterName }: Props) {
  const classes = useStyles()
  const modifiers = useSelector(selectAttributeModifiers(characterId))
  const { id, ...attributes } = useSelector(selectAttributes(characterId))

  function handleAttributeChange(key: keyof Attributes, value: string) {

  }

  if (!characterId) {
    return <Redirect to={{ pathname: '/' }} />
  }

  return (
    <context.Provider value={{ characterName, characterId }}>
      <div className={classes.characterSheet}>
        <InfoBar />
        <Attributes attributes={attributes} modifiers={modifiers} onChange={handleAttributeChange} />
        <CommonActivities />
        <EquipmentList />
        <Retainers />
        <SavingThrows />
        <AttackBonusAndHitPoints />
        <ArmorClassAndCombatOptions />
        <Encumbrance />
        <Languages />
      </div>
      <SpeedDial />
    </context.Provider>
  )
}

export {
  context as characterSheetContext
}

export default CharacterSheet