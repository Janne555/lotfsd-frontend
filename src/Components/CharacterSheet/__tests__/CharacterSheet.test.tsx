import React from 'react'
import { JssProvider } from 'react-jss'
import { render, waitForDomChange } from '@testing-library/react'
import { MockedProvider } from '@apollo/react-testing'
import { MemoryRouter } from 'react-router-dom'
import { LoginProvider } from '../../../hooks'
import { APOLLO_MOCKS } from '../../../testdata'
import CharacterSheet from '../CharacterSheet'

describe('<CharacterSheet />', () => {
  it('should match snapshot', async () => {
    const { asFragment } = render(
      <MemoryRouter>
        <LoginProvider>
          <MockedProvider mocks={APOLLO_MOCKS}>
            <JssProvider disableStylesGeneration>
              <CharacterSheet characterId="asd" characterName="Zoid" itemIndex={[]} />
            </JssProvider>
          </MockedProvider>
        </LoginProvider>
      </MemoryRouter>
    )

    await waitForDomChange()

    expect(asFragment()).toMatchSnapshot()
  });
});