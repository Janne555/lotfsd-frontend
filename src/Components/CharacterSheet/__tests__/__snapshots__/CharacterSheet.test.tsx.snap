// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CharacterSheet /> should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <h1>
        Lamentations of the fullstack developer
      </h1>
      <div>
        <div>
          <span>
            dyyd
          </span>
          <span>
            name
          </span>
        </div>
        <div>
          <span>
            0
          </span>
          <span>
            experience
          </span>
        </div>
        <div>
          <span>
            fighter
          </span>
          <span>
            class
          </span>
        </div>
        <div>
          <span>
            male
          </span>
          <span>
            race
          </span>
        </div>
        <div>
          <span>
            13
          </span>
          <span>
            age
          </span>
        </div>
        <div>
          <span>
            male
          </span>
          <span>
            gender
          </span>
        </div>
        <div>
          <span>
            chaotic
          </span>
          <span>
            alignment
          </span>
        </div>
      </div>
    </div>
    <div>
      <span>
        Ability Scores
      </span>
      <span>
        Modifiers
      </span>
      <label
        for="charisma"
      >
        Charisma
      </label>
      <div>
        <input
          id="charisma"
          type="number"
          value="10"
        />
      </div>
      <div>
        <span>
          0
        </span>
      </div>
      <span>
        Retainer Recruitment, Loyalty
      </span>
      <label
        for="constitution"
      >
        Constitution
      </label>
      <div>
        <input
          id="constitution"
          type="number"
          value="16"
        />
      </div>
      <div>
        <span>
          2
        </span>
      </div>
      <span>
        Hit Points, Daily Travel Distance
      </span>
      <label
        for="dexterity"
      >
        Dexterity
      </label>
      <div>
        <input
          id="dexterity"
          type="number"
          value="10"
        />
      </div>
      <div>
        <span>
          0
        </span>
      </div>
      <span>
        AC, Ranged AB, Initiative
      </span>
      <label
        for="intelligence"
      >
        Intelligence
      </label>
      <div>
        <input
          id="intelligence"
          type="number"
          value="11"
        />
      </div>
      <div>
        <span>
          0
        </span>
      </div>
      <span>
        Saves vs Magic Effects
      </span>
      <label
        for="strength"
      >
        Strength
      </label>
      <div>
        <input
          id="strength"
          type="number"
          value="11"
        />
      </div>
      <div>
        <span>
          0
        </span>
      </div>
      <span>
        Mêlée AB, Open Doors
      </span>
      <label
        for="wisdom"
      >
        Wisdom
      </label>
      <div>
        <input
          id="wisdom"
          type="number"
          value="7"
        />
      </div>
      <div>
        <span>
          -1
        </span>
      </div>
      <span>
        Saves vs Non-Magic Effects
      </span>
    </div>
    <div>
      <h2>
        Saving Throws
      </h2>
      <div>
        <div>
          <h3>
            Breath Weapon
          </h3>
          <div>
            <input
              value="11"
            />
          </div>
          <span>
            Area Effects
          </span>
        </div>
        <div>
          <h3>
            Magic
          </h3>
          <div>
            <input
              value="11"
            />
          </div>
          <span>
            Spells or Innate Abilities
          </span>
        </div>
        <div>
          <h3>
            Magical Device
          </h3>
          <div>
            <input
              value="12"
            />
          </div>
          <span>
            Spell-Like Effects from Items
          </span>
        </div>
        <div>
          <h3>
            Paralyze
          </h3>
          <div>
            <input
              value="14"
            />
          </div>
          <span>
            Mobility Hazards (Petrification, Hold, Etc.)
          </span>
        </div>
        <div>
          <h3>
            Poison
          </h3>
          <div>
            <input
              value="15"
            />
          </div>
          <span>
            Instant Death/KO Situations
          </span>
        </div>
      </div>
    </div>
    <div>
      <div>
        <h2>
          Attack Bonus
        </h2>
        <div>
          <div>
            <input
              value="1"
            />
          </div>
        </div>
        <h4>
          Base AB
        </h4>
        <div>
          <div>
            1
          </div>
        </div>
        <h4>
          Mêlée AB
        </h4>
        <div>
          <div>
            1
          </div>
        </div>
        <h4>
          Ranged AB
        </h4>
      </div>
      <div>
        <h3>
          Surprise Chance
        </h3>
        <div>
          <div>
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
      </div>
      <div>
        <h4>
          Max HP
        </h4>
        <div>
          <span>
            <input
              type="number"
              value="13"
            />
          </span>
        </div>
        <div>
          <span>
            <input
              type="number"
              value="13"
            />
          </span>
        </div>
        <h4>
          Current HP
        </h4>
      </div>
    </div>
    <div>
      <div>
        <h2>
          Armor Class
        </h2>
        <div>
          <div>
            <span>
              12
            </span>
          </div>
          <h4>
            Mêlée
          </h4>
          <div>
            <span>
              12
            </span>
          </div>
          <h4>
            Ranged
          </h4>
          <div>
            <span>
              12
            </span>
          </div>
          <h4>
            Without
            <br />
            Shield
          </h4>
          <div>
            <span>
              10
            </span>
          </div>
          <h4>
            Surprised
          </h4>
        </div>
      </div>
      <div>
        <h2>
          Combat Options
        </h2>
        <p>
          Standard attack AB +0, AC +0
        </p>
        <p>
          Parry AC 4
        </p>
        <p>
          Press AB +2, AC -4
        </p>
        <p>
          Defensive AB -4, AC +2
        </p>
      </div>
    </div>
    <div>
      <h2>
        Common Activities
      </h2>
      <div>
        <div>
          <h4>
            Architecture
          </h4>
          <div>
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
        <div>
          <h4>
            Bushcraft
          </h4>
          <div>
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
        <div>
          <h4>
            Climbing
          </h4>
          <div>
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
        <div>
          <h4>
            Languages
          </h4>
          <div>
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
        <div>
          <h4>
            Open Doors
          </h4>
          <div>
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
        <div>
          <h4>
            Search
          </h4>
          <div>
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
        <div>
          <h4>
            Sleight of Hand
          </h4>
          <div>
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
        <div>
          <h4>
            Sneak Attack
          </h4>
          <div>
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
        <div>
          <h4>
            Stealth
          </h4>
          <div>
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
        <div>
          <h4>
            Tinkering
          </h4>
          <div>
            <div />
            <div />
            <div />
            <div />
            <div />
            <div />
          </div>
        </div>
      </div>
    </div>
    <div>
      <h2>
        Equipment
      </h2>
      <table>
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Stack size
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Money
            </td>
            <td>
              100/100
            </td>
          </tr>
          <tr>
            <td>
              Money
            </td>
            <td>
              2/100
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div>
      <h2>
        Encumbrance
      </h2>
      <div>
        <h3>
          Total
        </h3>
        <div>
          0
        </div>
        <h4>
          Unencumbered
        </h4>
      </div>
      <h4>
        Explore
      </h4>
      <span>
        120 feet
      </span>
      <h4>
        Combat
      </h4>
      <span>
        40 feet
      </span>
      <h4>
        Running
      </h4>
      <span>
        120 feet
      </span>
      <h4>
        Per Day
      </h4>
      <span>
        24 miles
      </span>
    </div>
    <div>
      <div>
        <h2>
          Retainers
        </h2>
        <div>
          No Content
        </div>
      </div>
    </div>
    <div>
      <div>
        <h2>
          Languages
        </h2>
        <div>
          No Content
        </div>
      </div>
    </div>
    <div>
      <div>
        <h2>
          Properties
        </h2>
        <table>
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                Location
              </th>
              <th>
                Value
              </th>
              <th>
                Rent
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
